---
title: "Class 14: RNA-Seq analysis mini project"
author: "Dylan Mullaney (A16869792("
format: pdf
toc: true
---

## Background

The data for this session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only

## Data import

```{r}
#| message: false
library(DESeq2)
```

```{r}
counts <- read.csv("GSE37704_featurecounts.csv", row.names=1)
colData <-read.csv("GSE37704_metadata.csv")
```

## Inspect and tidy data

Does the `counts` columns match the `colData` rows?

```{r}
head(colData)
head(counts)
```

`Counts` has a lot of data going on. The first column, "length" does not match the `colData` values.

```{r}
colData$id
colnames(counts)
```

We need to remove the first column from counts to make it match up. Only run this once or it will keep taking away columns. 

```{r}
countData <- counts [, -1]
```

```{r}
colnames(countData) == colData$id
```

> Q1. How many genes in total?

```{r}
nrow(countData)
```

> Q2. Filter to remove zero count genes (rows where the gene activity is zero in all columns)

```{r}
#library(dplyr)
#countData %>% filter (rowSums(countData) == 0)

to.keep.inds <- rowSums(countData) >0
```

```{r}
new.counts<- countData[to.keep.inds,]
nrow(new.counts)
```

## Setup for DESeq

Already called the function in by library above

```{r}
dds <- DESeqDataSetFromMatrix(countData = new.counts,
                              colData = colData,
                              design= ~condition)
```


## Run DESeq

```{r}
dds <- DESeq(dds)
res <- results (dds)
```

```{r}
head(res)
```

## Volcano plot of results

```{r}
library(ggplot2)
```

```{r}
```


```{r}
ggplot(res) + 
       aes(log2FoldChange, -log(padj)) +
       geom_point() +
       geom_vline (xintercept= c(-2,2), col="orange")+
       geom_hline (yintercept = -log(0.05), col="red")
```

## Gene annotation

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=rownames(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=rownames(res),
                    keytype="ENSEMBL",
                    column="ENTREZID")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype= "ENSEMBL",
                    column="GENENAME")

head(res, 10)

```

## Pathway analysis

```{r}
library(gage)
library(gageData)
library(pathview)
```

Input vector for gage()

```{r}
foldchanges = res$log2FoldChange
names (foldchanges) =res$entrez
```

```{r}
data("kegg.sets.hs")
keggres = gage (foldchanges, gsets=kegg.sets.hs)
```

Cell cycle figure

```{r}
pathview(foldchanges, pathway.id= "hsa04110")
```

MAPK Signaling Pathway

![](hsa04110.pathview.png)

Let's look at a second pathway for Caffeine Metabolism

```{r}
kegg.sets.hs$`hsa00232 Caffeine metabolism`
```


```{r}
pathview(foldchanges, pathway.id= "hsa00232")
```

![](hsa00232.pathview.png)

Okay one last pathway for primary immunodeficiency

```{r}
pathview(foldchanges, pathway.id= "hsa05340")
```

![](hsa05340.pathview.png)

## Gene Ontology (GO)

Let's look at biological processes within gene ontology

```{r}
data(go.sets.hs)
data(go.subs.hs)

#biological processes subset
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

This divides into values greater than and values less than. I'm not sure exactly what I did here. 


## Reactome Analysis

Reactome is database consisting of biological molecules and their relation to pathways and processes (https://reactome.org/). 

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

> Q: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?

The "Cel Cycle" has the most significant 'entities P-value'. Most of these pathways seem to align with the previous predictions. The cell growth and programmed cell death section feel like they could counteract each other. 




